scripts/rollup.js

```diff
@@ -0,0 +1,178 @@
+import { MultiIndexWriter, MultiIndexReader } from 'cardex/multi-index'
+import { encodeVarint } from 'cardex/encoder'
+import * as Link from 'multiformats/link'
+import { base58btc } from 'multiformats/bases/base58'
+
+dotenv.config()
+
+async function main () {
+  if (!process.argv[2]) {
+    console.error('❌ missing root CID argument')
+    process.exit(1)
+  }
+  const root = Link.parse(process.argv[2])
+  console.log(`⏳ Generating rollup for ${root}...`)
+
+  const s3 = new S3Client({
+    region: notNully(process.env, 'S3_REGION'),
+    endpoint: process.env.S3_ENDPOINT,
+    credentials: {
+      accessKeyId: notNully(process.env, 'S3_ACCESS_KEY_ID'),
+      secretAccessKey: notNully(process.env, 'S3_SECRET_ACCESS_KEY')
+    }
+  })
+
+  const carparkBucket = notNully(process.env, 'CARPARK_BUCKET')
+  const dudewhereBucket = notNully(process.env, 'DUDEWHERE_BUCKET')
+  const satnavBucket = notNully(process.env, 'SATNAV_BUCKET')
+
+  const shardKeys = await listAll(s3, dudewhereBucket, `${root}/`)
+  /** @type {import('cardex/api').CARLink[]} */
+  const shards = []
+  for (const k of shardKeys) {
+    try {
+      shards.push(Link.parse(k.split('/').pop() ?? ''))
+    } catch (err) {
+      console.warn(`⚠️ failed to parse CID in key: ${k}`)
+    }
+  }
+
+  console.log(`ℹ️ DAG found in ${shards.length} shards`)
+
+  // Ensure shards exist in CARPARK
+  let shardNum = 0
+  for (const shard of shards) {
+    try {
+      const cmd = new HeadObjectCommand({ Bucket: carparkBucket, Key: `${shard}/${shard}.car` })
+      const res = await s3.send(cmd)
+      console.log(`✅ #${shardNum} Shard present: ${shard} (${res.ContentLength} bytes)`)
+      shardNum++
+    } catch (err) {
+      console.error(`❌ shard not found: ${shard}: ${err}`)
+      process.exit(1)
+    }
+  }
+
+  shardNum = 0
+  const readable = new ReadableStream({
+    start (controller) {
+      controller.enqueue(encodeVarint(MultiIndexWriter.codec))
+      controller.enqueue(encodeVarint(shards.length))
+    },
+    async pull (controller) {
+      const shard = shards[shardNum]
+      if (!shard) return controller.close()
+      controller.enqueue(shard.multihash.bytes)
+      const key = `${shard}/${shard}.car.idx`
+      const cmd = new GetObjectCommand({ Bucket: satnavBucket, Key: key })
+      const res = await s3.send(cmd)
+      if (!res.Body) throw new Error('missing index body')
+      const bytes = await res.Body.transformToByteArray()
+      controller.enqueue(bytes)
+      console.log(`✅ #${shardNum} Index read: ${shard}`)
+      shardNum++
+    }
+  })
+
+  const chunks = []
+  // @ts-expect-error
+  for await (const chunk of readable) {
+    chunks.push(chunk)
+  }
+  const index = Buffer.concat(chunks)
+
+  console.log(`⏳ Writing index (${index.length} bytes)...`)
+  const key = `${root}/.rollup.idx`
+  const cmd = new PutObjectCommand({ Bucket: dudewhereBucket, Key: key, Body: index })
```

Comment:
minor: Could convert your `ReadableStream` to a node `Readable` and pass that as `Body`

https://github.com/awslabs/smithy-typescript/blob/f785a574d75899c5fa91f8c5b4ac5bbae2564ed6/packages/types/src/streaming-payload/streaming-blob-payload-input-types.ts#L56C90-L56C98